src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -868,98 +1128,1238 @@ grpc_subchannel_call *grpc_client_channel_get_subchannel_call(
+  if (batch->send_message) return 1;
+  if (batch->send_trailing_metadata) return 2;
+  if (batch->recv_initial_metadata) return 3;
+  if (batch->recv_message) return 4;
+  if (batch->recv_trailing_metadata) return 5;
+  GPR_UNREACHABLE_CODE(return (size_t)-1);
+}
+
+// Cleans up retry state.  Called when the RPC is committed (i.e., we will
+// not attempt any more retries).
+static void retry_commit(grpc_exec_ctx *exec_ctx, grpc_call_element *elem,
+                         subchannel_call_retry_state *retry_state) {
+  call_data *calld = (call_data *)elem->call_data;
+  channel_data *chand = (channel_data *)elem->channel_data;
+  if (calld->retry_committed) return;
+  calld->retry_committed = true;
+  if (GRPC_TRACER_ON(grpc_client_channel_trace)) {
+    gpr_log(GPR_DEBUG, "chand=%p calld=%p: committing retries", chand, calld);
+  }
+  if (retry_state == NULL) return;
+  if (retry_state->completed_send_initial_metadata) {
+    grpc_metadata_batch_destroy(exec_ctx, &calld->send_initial_metadata);
+  }
+  for (size_t i = 0; i < retry_state->completed_send_message_count; ++i) {
+    grpc_byte_stream_cache_destroy(exec_ctx, calld->send_messages[i]);
+  }
+  if (retry_state->completed_send_trailing_metadata) {
+    grpc_metadata_batch_destroy(exec_ctx, &calld->send_trailing_metadata);
+  }
+}
+
 // This is called via the call combiner, so access to calld is synchronized.
-static void waiting_for_pick_batches_add(
-    call_data *calld, grpc_transport_stream_op_batch *batch) {
+static void pending_batches_add(grpc_exec_ctx *exec_ctx,
+                                grpc_call_element *elem,
+                                grpc_transport_stream_op_batch *batch) {
+  call_data *calld = (call_data *)elem->call_data;
+  channel_data *chand = (channel_data *)elem->channel_data;
+  const size_t idx = get_batch_index(batch);
+  if (GRPC_TRACER_ON(grpc_client_channel_trace)) {
+    gpr_log(GPR_DEBUG,
+            "chand=%p calld=%p: adding pending batch at index %" PRIuPTR, chand,
+            calld, idx);
+  }
+  pending_batch *pending = &calld->pending_batches[idx];
+  GPR_ASSERT(pending->batch == NULL);
+  pending->batch = batch;
+  pending->send_ops_cached = false;
   if (batch->send_initial_metadata) {
-    GPR_ASSERT(calld->initial_metadata_batch == NULL);
-    calld->initial_metadata_batch = batch;
-  } else {
-    GPR_ASSERT(calld->waiting_for_pick_batches_count < MAX_WAITING_BATCHES);
-    calld->waiting_for_pick_batches[calld->waiting_for_pick_batches_count++] =
-        batch;
+    calld->pending_send_initial_metadata = true;
+  }
+  if (batch->send_message) {
+    calld->pending_send_message = true;
+  }
+  if (batch->send_trailing_metadata) {
+    calld->pending_send_trailing_metadata = true;
+  }
+  // Check if the batch takes us over the retry buffer limit.
+  // Note: We don't check trailing metadata here, because gRPC clients
+  // do not send trailing metadata.
+  if (batch->send_initial_metadata) {
```

Comment:
Combine the bodies of these two conditionals with the earlier if (batch->send_initial_metadata) and if (batch->send_message)